lib/Dialect/HW/Transforms/LowerHWtoBTOR2.cpp

```diff
@@ -0,0 +1,786 @@
+      if (w == p.first) {
+        sortToLIDMap[i] = std::make_pair<size_t, size_t>(
+            std::forward<size_t>(w), std::forward<size_t>(lid));
+        return;
+      }
+    }
+
+    // Otherwise simply create a new entry
+    sortToLIDMap.push_back(std::make_pair<size_t, size_t>(
+        std::forward<size_t>(w), std::forward<size_t>(lid)));
+  }
+
+  // Checks if an operation was declared
+  // If so, its lid will be returned
+  // Otherwise -1 will be returned
+  size_t getOpLID(Operation *op) {
+    // Look for the operation declaration
+    Operation *defOp = getOpAlias(op);
+    if (opLIDMap.contains(defOp)) {
+      return opLIDMap[defOp];
+    }
+
+    // If no lid was found return -1
+    return NO_LID;
+  }
+
+  // Checks if an operation was declared
+  // If so, its lid will be returned
+  // Otherwise -1 will be returned
+  size_t getOpLID(Value op) {
+    // Look for the operation declaration
+    Operation *defOp = getOpAlias(op.getDefiningOp());
+
+    if (opLIDMap.contains(defOp)) {
+      return opLIDMap[defOp];
+    }
+
+    // Check for special case where op is actually a port
+    // To do so, we start by checking if our operation isa block argument
+    if (BlockArgument barg = dyn_cast<BlockArgument>(op)) {
+      // Extract the block argument index and use that to get the line number
+      size_t argIdx = barg.getArgNumber();
+
+      // Check that the extracted argument is in range
+      if (inputLIDs.contains(argIdx)) {
+        return inputLIDs[argIdx];
+      }
+    }
+
+    // If no lid was found return -1
+    return NO_LID;
+  }
+
+  // Updates or creates an entry for the given operation
+  // associating it with the current lid
+  void setOpLID(Operation *op) {
+    if (op != nullptr) {
+      opLIDMap[op] = lid;
+    }
+  }
+
+  // Checks if an operation has an alias
+  // If so, the original operation is returned
+  // Otherwise the argument is returned as it is the original op
+  Operation *getOpAlias(Operation *op) {
+    // Look for the operation declaration
+    if (opAliasMap.contains(op)) {
+      return opAliasMap[op];
+    }
+    // If no lid was found return -1
+    return op;
+  }
+
+  // Updates or creates an entry for the given operation
+  // associating it with the current lid
+  void setOpAlias(Operation *alias, Operation *op) { opAliasMap[alias] = op; }
+
+  /// String generation helper functions
+
+  // Generates a sort declaration instruction given a type (bitvec or array) and
+  // a width
+  void genSort(StringRef type, size_t width) {
+    // Check that the sort wasn't already declared
+    if (getSortLID(width) != NO_LID) {
+      return; // If it has already been declared then return an empty string
+    }
+
+    // Add the width to the declared sorts along with the associated line id
+    setSortLID(width);
+
+    // Build and return a sort declaration
+    llvm::outs() << lid++ << ws << sort << ws << bitvec << ws << width << nl;
```

Comment:
As mentioned, this isn't a way to generate output files.  You need to pass the stream to these functions and have the caller make a valid output file stream.

Reply:
I tried to do that now, not sure that I did it correctly but I based myself off of what I saw in other passes, given the pass a `raw_ostream` as an argument